### DISCUSSION TITLE
Guide on how to deploy as Azure static website or app service

### DISCUSSION THREAD

**Discussion by AsuraKev** [web-infra-dev/modern.js#4657] [Q&A]
Hi there, 

are there or will there be a in-depth deployment guide to some cloud infrastructure such as Azure? Also is it not possible to directly run the index.html on the browser after the build?

thanks

**Comment by chenjiahan** [comment]
> Also is it not possible to directly run the index.html on the browser after the build?

You can run the `modern serve` command to run production app: https://modernjs.dev/en/apis/app/commands.html#modern-serve

**Reply by AllenAttuned** [reply]
So there's no way to have it produce a build output that can just be served statically? Like, there's no method to produce a build that can be used with something like S3, `npx serve`, or standard web servers like Apache/Nginx? I personally think that one of the main benefits for SPAs is their ability to be served from practically anywhere without the need for a JavaScript runtime.

Also, the DX of this framework is amazing, and I love that you're treating SPAs with the same respect as SSR or SSG.

**Reply by xc2** [reply]
https://modernjs.dev/en/apis/app/commands.html#modern-build

`modern build` and the `dist` dir is what you're looking for

**Reply by AllenAttuned** [reply]
Thanks for linking the docs. So you can you just dump those static files somewhere and then serve them? Usually, what I do to test a build is use `npx serve` on the build directory. I missed it yesterday, but the build output names the "main" HTML file `main.html`. Renaming this to `index.html` seemed to solve the issue, and I can serve the static files using something like `npx serve`.

```
.
|____main.f6349dbe.js
|____lib-router.a2113bfd.js
|____476.dfb6852e.js.LICENSE.txt
|____async
| |____products
| | |____(id)
| | | |____page.79a8ae59.js
| | |____page.35c17bfa.js
| |____page.a619a68a.js
| |____101.a5464db7.js
| |____101.a5464db7.js.LICENSE.txt
| |____page.d7915515.css
|____lib-react.eb4f86f6.js
|____lib-react.eb4f86f6.js.LICENSE.txt
|____nestedRoutes.json
|____modern.config.json
|____route.json
|____main.html # <---- This seemed to be the issue
|____routes-manifest.json
|____lib-router.a2113bfd.js.LICENSE.txt
|____lib-polyfill.1a011935.js
|____476.dfb6852e.js
```

Is it possible to configure Modern.js to name the main point of entry `index.html` rather than `main.html`?

This is my `modern.config.js`

```javascript
import { appTools, defineConfig } from '@modern-js/app-tools';

// https://modernjs.dev/en/configure/app/usage
export default defineConfig({
  output: {
    distPath: {
      js: '',
      css: '',
      html: '',
    },
    cleanDistPath: true,
    disableSourceMap: true,
  },
  html: {
    disableHtmlFolder: true,
  },
  runtime: {
    router: true,
  },
  plugins: [
    appTools({
      bundler: 'webpack', // Set to 'experimental-rspack' to enable rspack ⚡️🦀
    }),
  ],
});
```

**Reply by xc2** [reply]
[source.mainEntryName](https://modernjs.dev/en/configure/app/source/mainEntryName.html) will help

**Reply by AllenAttuned** [reply]
Amazing, thank you so much!

**Reply by AllenAttuned** [reply]
Apologies, another question.

If you're deploying in this manner, not using `modern serve`, do you lose out on the routing features and have to use [Self-controlled routing](https://modernjs.dev/en/guides/basic-features/routes.html#self-controlled-routing)?

When I'm serving using `npx serve`, I can load the web app and routing works as expected if navigating using the `<Link>` elements in the UI. However, if I try and go to the route directly, or refresh the page while on a route, I'll get a `404`.

This is my route structure:

```
.
|____modern-app-env.d.ts
|____.eslintrc.js
|____routes
| |____products
| | |____page.data.ts
| | |____[id]
| | | |____page.data.ts
| | | |____page.tsx
| | |____page.tsx
| |____index.css
| |____layout.tsx
| |____page.tsx
```

Here's my config

```typescript
export default defineConfig({
  output: {
    distPath: {
      html: '',
    },
    cleanDistPath: true,
  },
  html: {
    disableHtmlFolder: true,
  },
  source: {
    mainEntryName: 'index'
  },
  runtime: {
    router: {
      supportHtml5History: true,
    }
  },
  plugins: [
    appTools({
      bundler: 'webpack',
    }),
  ],
});
```

**Reply by xc2** [reply]
further discussion moved to https://github.com/web-infra-dev/modern.js/discussions/5501

**Comment by chenjiahan** [comment]
> are there or will there be a in-depth deployment guide to some cloud infrastructure such as Azure?

I think we will work on it, related issue: https://github.com/web-infra-dev/modern.js/issues/3624

If you are interested in contributing relevant documentation, feel free to send a pull request 😄
